File: test-suite/src/test/java/io/micronaut/microstream/docs/CustomerRepositoryStoreImpl.java
package io.micronaut.microstream.docs;

import io.micronaut.context.annotation.Requires;
import io.micronaut.core.annotation.NonNull;
import io.micronaut.core.annotation.Nullable;
import io.micronaut.microstream.annotation.Store;
import jakarta.inject.Singleton;
import one.microstream.storage.embedded.types.EmbeddedStorageManager;

import javax.validation.Valid;
import javax.validation.constraints.NotBlank;
import javax.validation.constraints.NotNull;
import java.util.Map;
import java.util.Optional;
import java.util.UUID;

@Requires(property = "customer.repository", value = "store")
//tag::clazz[]
@Singleton
public class CustomerRepositoryStoreImpl implements CustomerRepository {

    private final EmbeddedStorageManager embeddedStorageManager;

    public CustomerRepositoryStoreImpl(EmbeddedStorageManager embeddedStorageManager) { // <1>
        this.embeddedStorageManager = embeddedStorageManager;
    }

    @Override
    @NonNull
    public Customer save(@NonNull @NotNull @Valid CustomerSave customerSave) {
        return addCustomer(data().getCustomers(), customerSave);
    }

    @Override
    public void update(@NonNull @NotBlank String id,
                       @NonNull @NotNull @Valid CustomerSave customerSave) {
        updateCustomer(id, customerSave);
    }

    @Override
    @NonNull
    public Optional<Customer> findById(@NonNull @NotBlank String id) {
        return Optional.ofNullable(data().getCustomers().get(id));
    }

    @Override
    public void deleteById(@NonNull @NotBlank String id) {
        removeCustomer(data().getCustomers(), id);
    }

    @Store(result = true) // <2>
    @Nullable
    protected Customer updateCustomer(@NonNull String id,
                                      @NonNull CustomerSave customerSave) {
        Customer c = data().getCustomers().get(id);
        if (c != null) {
            c.setFirstName(customerSave.getFirstName());
            c.setLastName(customerSave.getLastName());
            return c;
        }
        return null;
    }

    @Store(parameters = "customers") // <3>
    protected Customer addCustomer(@NonNull Map<String, Customer> customers,
                                   @NonNull CustomerSave customerSave) {
        Customer customer = new Customer(UUID.randomUUID().toString(),
            customerSave.getFirstName(),
            customerSave.getLastName());
        customers.put(customer.getId(), customer);
        return customer;
    }

    @Store(parameters = "customers") // <3>
    protected void removeCustomer(@NonNull Map<String, Customer> customers,
                                  @NonNull String id) {
        customers.remove(id);
    }

    @NonNull
    private Data data() {
        return (Data) embeddedStorageManager.root();
    }
}
//end::clazz[]

File: test-suite-python/src/test/python/micronaut/microstream/docs/CustomerRepositoryStoreImpl.py
import uuid
from typing import Annotated

from jakarta.inject import Singleton
from jakarta.validation import Valid
from jakarta.validation.constraints import NotBlank, NotNull
from micronaut.context.annotation import Requires
from micronaut.microstream import RootProvider
from micronaut.microstream.annotations import StoreParams, StoreReturn

from .Customer import Customer
from .CustomerRepository import CustomerRepository
from .CustomerSave import CustomerSave
from .Data import Data


@Requires(property="customer.repository", value="store")
# tag::clazz[]
@Singleton
class CustomerRepositoryStoreImpl(CustomerRepository):

    def __init__(self, root_provider: RootProvider[Data]):  # <1>
        self.root_provider = root_provider

    def save(self, customer_save: Annotated[CustomerSave, NotNull, Valid]) -> Customer:
        return self.add_customer(self.root_provider.root().customers, customer_save)

    def update(self, id: Annotated[str, NotBlank],
               customer_save: Annotated[CustomerSave, NotNull, Valid]) -> None:
        self.update_customer(id, customer_save)

    def find_by_id(self, id: Annotated[str, NotBlank]) -> Customer | None:
        return self.root_provider.root().customers.get(id)

    def delete_by_id(self, id: Annotated[str, NotBlank]) -> None:
        self.remove_customer(self.root_provider.root().customers, id)

    @StoreReturn  # <2>
    def update_customer(self, id: str, customer_save: CustomerSave) -> Customer | None:
        c = self.root_provider.root().customers.get(id)
        if c is not None:
            c.firstName = customer_save.firstName
            c.lastName = customer_save.lastName
            return c
        return None

    @StoreParams("customers")  # <3>
    def add_customer(self, customers: dict[str, Customer], customer_save: CustomerSave) -> Customer:
        customer = Customer(str(uuid.uuid4()), customer_save.firstName, customer_save.lastName)
        customers[customer.id] = customer
        return customer

    @StoreParams("customers")  # <3>
    def remove_customer(self, customers: dict[str, Customer], id: str) -> None:
        customers.pop(id, None)
# end::clazz[]
